### ISSUE TITLE
实验求助

### ISSUE THREAD

**Issue by fahai-dd** [demonph10/MGCN#6]
作者您好：
请问您关于论文中单模态的实验是如何做的？行为融合器的代码应该如何处理（只剩下一个模态了），能否详细说明一下 非常感谢！祝生活顺利

**Comment by demonph10** [comment]
由于行为融合器原理上类似于attention，因此在仅剩一个模态时候，我们是将行为信息和模态信息进行了attention的计算作为新的模态偏好信息，再和行为信息相加。

**Comment by Gs0098** [comment]
非常感谢您的回复，关于这个实验我还想请问您，做单模态时，只修改融合器部分就可以了嘛，我进行了几次实验，感觉自己改的不太对，代码如下：        if self.modality_type == 'both':
            # --- 原始双模态逻辑 ---
            att_common = torch.cat([self.query_common(image_embeds), self.query_common(text_embeds)], dim=-1)
            weight_common = self.softmax(att_common)
            common_embeds = weight_common[:, 0].unsqueeze(dim=1) * image_embeds + weight_common[:, 1].unsqueeze(dim=1) * text_embeds
            sep_image_embeds = image_embeds - common_embeds
            sep_text_embeds = text_embeds - common_embeds
            image_prefer = self.gate_image_prefer(content_embeds)
            text_prefer = self.gate_text_prefer(content_embeds)
            sep_image_embeds = torch.multiply(image_prefer, sep_image_embeds)
            sep_text_embeds = torch.multiply(text_prefer, sep_text_embeds)
            side_embeds = (sep_image_embeds + sep_text_embeds + common_embeds) / 3

        elif self.modality_type == 'image':
            # --- 单模态 (仅图像) 逻辑 ---
            if image_embeds is not None:
                image_prefer = self.gate_image_prefer(content_embeds)
                side_embeds = torch.multiply(image_prefer, image_embeds)
            else: # 如果图像特征不存在，则side_embeds为0
                side_embeds = torch.zeros_like(content_embeds)

        elif self.modality_type == 'text':
            # --- 单模态 (仅文本) 逻辑 ---
            if text_embeds is not None:
                text_prefer = self.gate_text_prefer(content_embeds)
                side_embeds = torch.multiply(text_prefer, text_embeds)
            else: # 如果文本特征不存在，则side_embeds为0
                side_embeds = torch.zeros_like(content_embeds)

        else:
            # 如果配置错误，或者想进行纯ID实验，可以将side_embeds设为0
            print("Warning: modality_type is not 'both', 'image', or 'text'. Running ID-based only.")
            side_embeds = torch.zeros_like(content_embeds)

        # 最终表征 = 行为信息 + 模态偏好信息
        all_embeds = content_embeds + side_embeds
希望能得到您的指点！！万分感谢 祝您学业顺利，家人身体健康。

**Comment by demonph10** [comment]
> 非常感谢您的回复，关于这个实验我还想请问您，做单模态时，只修改融合器部分就可以了嘛，我进行了几次实验，感觉自己改的不太对，代码如下： if self.modality_type == 'both': # --- 原始双模态逻辑 --- att_common = torch.cat([self.query_common(image_embeds), self.query_common(text_embeds)], dim=-1) weight_common = self.softmax(att_common) common_embeds = weight_common[:, 0].unsqueeze(dim=1) * image_embeds + weight_common[:, 1].unsqueeze(dim=1) * text_embeds sep_image_embeds = image_embeds - common_embeds sep_text_embeds = text_embeds - common_embeds image_prefer = self.gate_image_prefer(content_embeds) text_prefer = self.gate_text_prefer(content_embeds) sep_image_embeds = torch.multiply(image_prefer, sep_image_embeds) sep_text_embeds = torch.multiply(text_prefer, sep_text_embeds) side_embeds = (sep_image_embeds + sep_text_embeds + common_embeds) / 3
> 
> ```
>     elif self.modality_type == 'image':
>         # --- 单模态 (仅图像) 逻辑 ---
>         if image_embeds is not None:
>             image_prefer = self.gate_image_prefer(content_embeds)
>             side_embeds = torch.multiply(image_prefer, image_embeds)
>         else: # 如果图像特征不存在，则side_embeds为0
>             side_embeds = torch.zeros_like(content_embeds)
> 
>     elif self.modality_type == 'text':
>         # --- 单模态 (仅文本) 逻辑 ---
>         if text_embeds is not None:
>             text_prefer = self.gate_text_prefer(content_embeds)
>             side_embeds = torch.multiply(text_prefer, text_embeds)
>         else: # 如果文本特征不存在，则side_embeds为0
>             side_embeds = torch.zeros_like(content_embeds)
> 
>     else:
>         # 如果配置错误，或者想进行纯ID实验，可以将side_embeds设为0
>         print("Warning: modality_type is not 'both', 'image', or 'text'. Running ID-based only.")
>         side_embeds = torch.zeros_like(content_embeds)
> 
>     # 最终表征 = 行为信息 + 模态偏好信息
>     all_embeds = content_embeds + side_embeds
> ```
> 
> 希望能得到您的指点！！万分感谢 祝您学业顺利，家人身体健康。

您好，当模态不存在时，不应该初始化为0张量，因为0也算是特征表示。如之前所提到的：

> 由于行为融合器原理上类似于attention，因此在仅剩一个模态时候，我们是将行为信息和模态信息进行了attention的计算作为新的模态偏好信息，再和行为信息相加。

当仅剩一个模态时，我们首先消除对应模态的特征净化器，然后用类似attention机制的计算来进行最终的不同模态的偏好融合。
